typing/typecore.ml

```diff
@@ -975,11 +975,18 @@ let solve_Ppat_record_field ~refine loc penv label label_lid record_ty =
   end
 
 let solve_Ppat_array ~refine loc env expected_ty =
-  let ty_elt = newgenvar() in
   let expected_ty = generic_instance expected_ty in
-  unify_pat_types_refine ~refine
-    loc env (Predef.type_array ty_elt) expected_ty;
-  ty_elt
+  if is_principal expected_ty &&
+     match get_desc (expand_head !!env expected_ty) with
+     | Tconstr (path, [], _) -> Path.same path Predef.path_floatarray
+     | _ -> false
+  then
+    Predef.type_float, Predef.type_floatarray
```

Comment:
Someone more expert in these matters than me may say otherwise, but I think you want `generic_instance Predef.type_float` for the element type. This seems to match the fact that the general case calls `newgenvar`.

Generally speaking, you don't want to use types from `Predef` directly; you usually want to take an `instance` or `generic_instance` of them. Using the type directly (without `instance`) means taking the master copy and putting it in a structure. If we then learn more about the type (say by a GADT pattern-match), we might influence that master copy.

Reply:
I ended up using `instance Predef.type_float` (I guess that for parameter-less types there is no difference). This matches what is done elsewhere in `Typecore`, eg `type_constant`.

Reply:
It surprised me, too, but there is a difference, actually. Both `instance` and `generic_instance` copy the type. `instance` additionally changes all levels that are `generic_level` to be the current level (that is, the result of `get_current_level ()`). We consider a type whose level is `generic_level` to be principally known: either this type came from a user-written type annotation, or inference on the type has been completed (for the type of a let-bound `x`) and the type has been generalized. So `instance Predef.type_float` produces a `float` that is not principally known (and couldn't be used in type-directed disambiguation, say) while `generic_instance Predef.type_float` *can* be used for type-directed disambiguation.

I see `Typecore.type_constant` uses `instance`. I'm not actually sure why that is: If I see that something is `1`, its type is definitely `int`; saying `(1 : int)` shouldn't be different from `1`. So I'm not exactly sure what's best here; we should wait for @garrigue.

Reply:
OK, I left `instance` for now until the issue is clarified. Thanks for the explanation by the way, it is appreciated!

Reply:
@goldfirere Type synthesized by `type_exp` are supposed to have a level no higher than `current_level`, which reflects the fact than syntax-directed rules for Hindley-Milner have only monotypes on the right-hand side. To make easier to follow this invariant, most functions return type nodes at `current_level`. Of course, one could start from `generic_level` and then use `update_level` to get back to `current_level`, but this introduces extra costs.

Reply:
I see what you're saying here, but the existence of `generic_level` is additionally used to affect principality of type inference. If I have an expression `5`, I would expect its type `int` to be principally known. This would suggest using `generic_instance` instead of `instance`. But now that I think about it, I can't find a way in which this would change any user-facing behavior, so maybe this isn't a point that needs debating...